src/server/pkg/ppsutil/decoder.go

```diff
@@ -0,0 +1,204 @@
+package ppsutil
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"unicode"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gogo/protobuf/proto"
+
+	ppsclient "github.com/pachyderm/pachyderm/src/client/pps"
+	"gopkg.in/pachyderm/yaml.v3"
+)
+
+type pipelineDecoder interface {
+	// Decode parses the next CreatePipelineRequest in the pipelineDecoder's input
+	// stream into 'v'
+	Decode(v interface{}) error
```

Comment:
Is there a reason that `v` can't be of type `*CreatePipelineRequest` it seems like that's the only type this is supposed to support.

Reply:
This actually also calls `Decode(map[string]interface{})` (see e.g. line 126 below). Basically by decoding it into a map, we can decode the TFJob in a structured way, and then serialize the TFJob and put it back before deserializing the whole request again into a struct that expects TFJob to be a string.